ggml-cuda.cu

```diff
@@ -1960,20 +1960,73 @@ static void ggml_cuda_mul_mat(ggml_backend_cuda_context & ctx, const ggml_tensor
     }
 }
 
+struct mmid_row_mapping {
+    int32_t i1;
+    int32_t i2;
+};
+
+static __global__ void k_copy_src1_to_contiguous(const char * __restrict__ src1_original, char * __restrict__ src1_contiguous,
+                                                 int * __restrict__ cur_src1_row, mmid_row_mapping * __restrict__ row_mapping,
+                                                 const char * __restrict ids, int64_t i02, size_t ids_nb1, size_t ids_nb0,
+                                                 int64_t ne11, int64_t ne10,
+                                                 size_t nb11, size_t nb12) {
+    int32_t iid1 = blockIdx.x;
+    int32_t id = blockIdx.y;
+
+    const int32_t row_id_i = *(const int32_t *) (ids + iid1*ids_nb1 + id*ids_nb0);
+
+    if (row_id_i != i02) {
+        return;
+    }
+
+    const int64_t i11 = id % ne11;
+    const int64_t i12 = iid1;
+
+    __shared__ int src1_row;
+    if (threadIdx.x == 0) {
+        src1_row = atomicAdd(cur_src1_row, 1);
```

Comment:
Would this `atomicAdd` make the results non-deterministic?

Reply:
The order of the src1 rows is not deterministic, so it is possible.

Reply:
I think we can accept if it is too difficult to avoid or if the performance would suffer significantly

Reply:
It could be fixed if it turns out to be a problem, but I tried a few generations with the same seed and they were identical.

Reply:
I don't think this is an issue for determinism. Usually you start to get issues with determinism if you start to do atomic adds with floats because there the order of operations matters for the final result. With integers additions are fully commutative (except for overflows) so to my understanding the only thing that is not deterministic here is which warp gets which specific task to work on.